src/util/provenanceUtils.js

```diff
@@ -0,0 +1,91 @@
+const { resolveSchema } = require('@projecttacoma/node-fhir-server-core');
+const { v4: uuidv4 } = require('uuid');
+
+/**
+ * Creates a raw JSON AuditEvent resource from the X-Provenance headers of a submission request
+ * @param {*} provenance the provenance headers in string form from the request
+ * @param {*} args the object containing the base_version parameter
+ * @returns A JSON object representing an AuditEvent to be stored in the system
+ */
+const createAuditEventFromProvenance = (provenance, args) => {
+  provenance = JSON.parse(provenance);
+  const audit = {};
+  audit.type = {
+    system: 'http://dicom.nema.org/resources/ontology/DCM',
+    code: '110100',
+    display: 'Application Activity'
+  };
+  audit.period = provenance.occurredPeriod;
+  audit.recorded = provenance.recorded;
+  audit.outcome = 0;
+  audit.purposeOfEvent = [];
+  if (provenance.reason) {
+    audit.purposeOfEvent.push(...provenance.reason);
+  }
+  switch (provenance?.activity?.coding?.[0]?.code) {
+    case 'UPDATE':
+      audit.action = 'U';
+      break;
+    default:
+      audit.action = 'C';
+      break;
+  }
+  let source = {};
+  audit.agent = [];
```

Comment:
According to the IG, there must be at least one audit agent and the agent must have a requestor. Will there ever be a case where the provenance does not have the `agent` attribute? In that case, I don't think the `audit.agent` will get a requestor because it won't go through the if statement `if (provenance.agent)`

Reply:
I think this will kind of depend. Since `agent` is `1..*` in provenance as well, theoretically we don't have to worry about Provenance missing it for the case of AuditEvent creation. However, it's TBD still on what we want to do if we receive an X-Provenance that isn't valid. Do we still create an AuditEvent that says that a submission failed because of an invalid Provenance? Maybe? So I think future tasking could be to add provenance validation to at least ensure that the required properties are there. If they're not, maybe we create an AuditEvent for a failed submission and can update the outcome accordingly.

For this task, I think it's fine to just assume that the provenance you're getting is good for a proper mapping, and we can worry about enforcing stuff later.

Reply:
EDIT: I see we have a helper function for validating provenance down below. If you want to just quickly enforce the 1..*-ness down there, that's cool. I think the future tasking that I'm talking about would only consider if we still want to create AuditEvents for bad provenance input, which can be scoped to a diff task